=============================================================
  AIRBNB MARKET RADAR — Step 0: Analisi Esplorativa Dataset
=============================================================

In [1]:
# import necessary libraries
import pandas as pd 
import os
import numpy as np

c:\Users\emanu\anaconda3\Lib\site-packages\pandas\core\computation\expressions.py:23: UserWarning: Pandas requires version '2.10.2' or newer of 'numexpr' (version '2.8.7' currently installed).
  from pandas.core.computation.check import NUMEXPR_INSTALLED
c:\Users\emanu\anaconda3\Lib\site-packages\pandas\core\arrays\masked.py:56: UserWarning: Pandas requires version '1.4.2' or newer of 'bottleneck' (version '1.3.7' currently installed).
  from pandas.core import (


In [2]:
# Define the file path to the CSV file
FILE_PATH = "Data/listings.csv"

In [3]:
# Load the CSV file into a pandas DataFrame
df = pd.read_csv(FILE_PATH)

In [4]:
# Dimensions of the DataFrame
rows, columns = df.shape
print(f"Apartaments: {rows:,}")
print(f"Columns (features): {columns}")
print()

Apartaments: 37,652
Columns (features): 18



In [5]:
# Name of the columns
for i, col in enumerate(df.columns):
    print(f"  [{i:>3}]  {col}")
print()

  [  0]  id
  [  1]  name
  [  2]  host_id
  [  3]  host_name
  [  4]  neighbourhood_group
  [  5]  neighbourhood
  [  6]  latitude
  [  7]  longitude
  [  8]  room_type
  [  9]  price
  [ 10]  minimum_nights
  [ 11]  number_of_reviews
  [ 12]  last_review
  [ 13]  reviews_per_month
  [ 14]  calculated_host_listings_count
  [ 15]  availability_365
  [ 16]  number_of_reviews_ltm
  [ 17]  license



In [6]:
# Identify the columns for our project
columns_of_interest = [
    "id",
    "neighbourhood",
    "latitude",
    "longitude",
    "room_type",
    "price",
]

In [7]:
# ==============================================
# Price analysis
# ==============================================

# Statistical summary of the 'price' column
price_stats = df["price"].describe()
print("Price Statistics:")
print(price_stats)

# First 10 rows of the 'price' column
print("\nFirst 10 rows of the 'price' column:")
print(df["price"].head(10).to_string())

# Check for missing values in the 'price' column
missing_price_count = df["price"].isnull().sum()
print(f"\nNumber of missing values in 'price' column: {missing_price_count}")

# Percentage of missing values in the 'price' column
total_rows = len(df)
missing_price_percentage = (missing_price_count / total_rows) * 100
print(f"Percentage of missing values in 'price' column: {missing_price_percentage:.2}")

# sample of non-null values in the 'price' column
print("\nSample of non-null values in the 'price' column:")
print(df["price"].dropna().head(5).tolist())

# check for duplicates in the 'id' column
duplicate_id_count = df["id"].duplicated().sum()
print(f"\nNumber of duplicate values in 'id' column: {duplicate_id_count}")

Price Statistics:
count    33564.000000
mean       197.328000
std        443.190521
min          8.000000
25%         93.000000
50%        133.000000
75%        195.250000
max      10515.000000
Name: price, dtype: float64

First 10 rows of the 'price' column:
0     57.0
1    110.0
2    124.0
3    162.0
4    150.0
5    195.0
6    245.0
7     95.0
8      NaN
9    136.0

Number of missing values in 'price' column: 4088
Percentage of missing values in 'price' column: 1.1e+01

Sample of non-null values in the 'price' column:
[57.0, 110.0, 124.0, 162.0, 150.0]

Number of duplicate values in 'id' column: 0


In [8]:
# ==============================================
# Latitude and Longitude analysis
# ==============================================

# First 10 rows of the 'latitude' and 'longitude' columns
print("\nFirst 10 rows of the 'latitude' column:")
print(df["latitude"].head(10).to_string())
print("\nFirst 10 rows of the 'longitude' column:")
print(df["longitude"].head(10).to_string())

# Check for missing values in the 'latitude' and 'longitude' columns
missing_latitude_count = df["latitude"].isnull().sum()
missing_longitude_count = df["longitude"].isnull().sum()
print(f"\nNumber of missing values in 'latitude' column: {missing_latitude_count}")
print(f"Number of missing values in 'longitude' column: {missing_longitude_count}")

# Percentage of missing values in the 'latitude' and 'longitude' columns
missing_latitude_percentage = (missing_latitude_count / total_rows) * 100
missing_longitude_percentage = (missing_longitude_count / total_rows) * 100
print(f"Percentage of missing values in 'latitude' column: {missing_latitude_percentage:.2f}%")
print(f"Percentage of missing values in 'longitude' column: {missing_longitude_percentage:.2f}%")

# sample of non-null values in the 'latitude' and 'longitude' columns
print("\nSample of non-null values in the 'latitude' column:")
print(df["latitude"].dropna().head(5).tolist())
print("\nSample of non-null values in the 'longitude' column:")
print(df["longitude"].dropna().head(5).tolist())


First 10 rows of the 'latitude' column:
0    41.871360
1    41.895447
2    41.925820
3    41.908230
4    41.908283
5    41.889920
6    41.889920
7    41.892940
8    41.910290
9    41.890840

First 10 rows of the 'longitude' column:
0    12.482150
1    12.491181
2    12.469280
3    12.452930
4    12.452617
5    12.468230
6    12.468230
7    12.505810
8    12.447030
9    12.468040

Number of missing values in 'latitude' column: 0
Number of missing values in 'longitude' column: 0
Percentage of missing values in 'latitude' column: 0.00%
Percentage of missing values in 'longitude' column: 0.00%

Sample of non-null values in the 'latitude' column:
[41.87136, 41.89544677734375, 41.92582, 41.90823, 41.90828323364258]

Sample of non-null values in the 'longitude' column:
[12.48215, 12.491181373596191, 12.46928, 12.45293, 12.452616691589355]


In [9]:
# ==============================================
# Room type analysis
# ==============================================

# Count of each unique value in the 'room_type' column
room_type_counts = df["room_type"].value_counts()
print("Room Type Counts:")
print(room_type_counts)

# Percentage of each room type
room_type_percentages = df["room_type"].value_counts(normalize=True) * 100
print("\nRoom Type Percentages:")
print(room_type_percentages)

# Check for missing values in the 'room_type' column
missing_room_type_count = df["room_type"].isnull().sum()
print(f"\nNumber of missing values in 'room_type' column: {missing_room_type_count}")

Room Type Counts:
room_type
Entire home/apt    28294
Private room        8887
Hotel room           370
Shared room          101
Name: count, dtype: int64

Room Type Percentages:
room_type
Entire home/apt    75.146075
Private room       23.602996
Hotel room          0.982684
Shared room         0.268246
Name: proportion, dtype: float64

Number of missing values in 'room_type' column: 0


In [10]:
# ==============================================
# Neighbourhood analysis
# ==============================================

# Count of each unique value in the 'neighbourhood' column
neighbourhood_counts = df["neighbourhood"].value_counts()
print("Neighbourhood Counts:")
print(neighbourhood_counts)

# Percentage of each neighbourhood
neighbourhood_percentages = df["neighbourhood"].value_counts(normalize=True) * 100
print("\nNeighbourhood Percentages:")
print(neighbourhood_percentages)

# Check for missing values in the 'neighbourhood' column
missing_neighbourhood_count = df["neighbourhood"].isnull().sum()
print(f"\nNumber of missing values in 'neighbourhood' column: {missing_neighbourhood_count}")

Neighbourhood Counts:
neighbourhood
I Centro Storico              18563
VII San Giovanni/Cinecittà     3264
XIII Aurelia                   2674
II Parioli/Nomentano           2619
XII Monte Verde                2014
V Prenestino/Centocelle        1888
VIII Appia Antica              1217
X Ostia/Acilia                 1049
XIV Monte Mario                 864
XI Arvalia/Portuense            781
IV Tiburtina                    683
III Monte Sacro                 669
XV Cassia/Flaminia              602
IX Eur                          424
VI Roma delle Torri             341
Name: count, dtype: int64

Neighbourhood Percentages:
neighbourhood
I Centro Storico              49.301498
VII San Giovanni/Cinecittà     8.668862
XIII Aurelia                   7.101880
II Parioli/Nomentano           6.955806
XII Monte Verde                5.348985
V Prenestino/Centocelle        5.014342
VIII Appia Antica              3.232232
X Ostia/Acilia                 2.786041
XIV Monte Mario                2.294

# Stage 0 — Data Preprocessing & Stage A — Hash Map Construction

## Objective
The pipeline receives a raw Airbnb dataset (37,652 listings, 18 columns) and must
suggest an optimal nightly price for a new host based on real competitor data.

## Stage 0: Data Preprocessing — O(N)
Before building any data structure, we perform a linear scan to remove listings
with missing or invalid prices. This prevents corrupted data from propagating
through the pipeline and causing incorrect results downstream.

**Edge cases handled:**
- Missing price (NaN): 4,088 listings removed (~11% of dataset)
- Zero or negative price: logically invalid, removed
- Result: ~33,564 clean listings ready for indexing

## Stage A: Hash Map Construction — Build O(N), Lookup Θ(1)
We index all clean listings into a Python dictionary using a composite key:
    key = "{room_type}_{neighbourhood}"

This gives us ~60 distinct categories (4 room types × 15 neighbourhoods).

**Why a Hash Map and not a Balanced BST?**
We do not need listings sorted in any order. We only need to extract
a specific category as fast as possible. A Hash Map guarantees Θ(1)
average-case lookup, beating the O(log n) of a BST.

**Memory design choice:**
We perform column selection during construction — no intermediate
DataFrame is created. Each record stored in the map contains only
the 4 fields needed by downstream stages: id, latitude, longitude, price.
The remaining 14 columns are never loaded into the dictionary.

In [11]:
import pandas as pd

# ─────────────────────────────────────────────────────────────
# STAGE 0 — Data Preprocessing
# Time complexity:  O(N) — single linear scan
# Space complexity: O(N) — cleaned dataframe in memory
# ─────────────────────────────────────────────────────────────

df = pd.read_csv(FILE_PATH, low_memory=False)

# Remove listings with missing or invalid prices
df = df[df["price"].notna()]  # removes 4,088 NaN entries (~11%)
df = df[df["price"] > 0]      # removes logically invalid prices

print(f"[Stage 0] Listings after cleaning: {len(df):,} / 37,652")


# ─────────────────────────────────────────────────────────────
# STAGE A — Hash Map Construction
# Time complexity build:  O(N) — single pass over cleaned data
# Time complexity lookup: Θ(1) — direct key access (avg case)
# Space complexity:       O(N) — one record per listing
# ─────────────────────────────────────────────────────────────

hash_map = {}  # key: "room_type_neighbourhood" → value: list of records

for _, row in df.iterrows():

    # Composite key — defines the competitor category
    key = f"{row['room_type']}_{row['neighbourhood']}"

    # Store only the 4 fields needed by downstream stages
    # Column selection happens here — no intermediate DataFrame created
    record = {
        "id":        row["id"],
        "latitude":  row["latitude"],
        "longitude": row["longitude"],
        "price":     row["price"],
    }

    # Θ(1) key existence check — property of Python dict (hash table)
    if key not in hash_map:
        hash_map[key] = []

    hash_map[key].append(record)

print(f"[Stage A] Hash Map built: {len(hash_map)} unique categories")
print(f"[Stage A] Sample keys: {list(hash_map.keys())[:4]}")

[Stage 0] Listings after cleaning: 33,564 / 37,652
[Stage A] Hash Map built: 50 unique categories
[Stage A] Sample keys: ['Private room_VIII Appia Antica', 'Entire home/apt_I Centro Storico', 'Entire home/apt_II Parioli/Nomentano', 'Private room_I Centro Storico']


## Stage B — Spatial Filter: KD-Tree Construction & KNN Query

## Objective
From the ~N listings returned by the Hash Map (Stage A), we need only the
K=50 geographically closest competitors to the host's apartment.
Comparing with the entire category would include listings in different 
neighbourhoods of Rome, which are not real competitors.

## KD-Tree Construction — O(N log N)
We build a KD-Tree on the (latitude, longitude) coordinates of all listings
in the selected category. The tree recursively splits the 2D geographic space
using alternating median cuts (~log₂N levels), creating a balanced binary tree.

This is the system BOTTLENECK — the most expensive single operation.
It is justified because it is paid once per query, and enables fast KNN search.

## KNN Query — Θ(log N + K)
Given the host's coordinates, the algorithm traverses the tree pruning branches
whose minimum possible distance to the host is already greater than the current
K-th closest distance found. In 2D (our case) pruning is highly effective,
visiting only O(log N + K) nodes instead of all N.

**Why KD-Tree and not Brute Force?**
Brute force computes distance to every listing: Θ(N).
KD-Tree prunes geometrically irrelevant regions: Θ(log N + K).
For N=14,000 (largest category): log₂(14,000) ≈ 14 vs 14,000 comparisons.

**Why not ANN (Approximate Nearest Neighbors)?**
ANN sacrifices exactness for speed in high dimensions (d >> 2).
Our space is exactly 2D (lat/lon) — the curse of dimensionality does not apply.
KD-Tree gives exact results with logarithmic query time. ANN is unnecessary.

**Edge cases handled:**
- Small categories (e.g. Shared room): K = min(50, N//2) prevents requesting
  more neighbors than available listings
- Duplicate coordinates (same building, different hosts): treated as distinct
  points, both included — correct since they have independent prices

In [12]:
import numpy as np
from scipy.spatial import KDTree

# ─────────────────────────────────────────────────────────────
# STAGE B — KD-Tree Construction + KNN Query
# Construction time:  O(N log N) — SYSTEM BOTTLENECK
# Query time:         Θ(log N + K)
# Space complexity:   O(N) — stores all N coordinate pairs
# ─────────────────────────────────────────────────────────────

def build_and_query_kdtree(listings, host_lat, host_lon):
    """
    Build a KD-Tree on competitor coordinates and find K nearest neighbors.
    
    Parameters:
        listings  : list of dicts from Hash Map — each has lat, lon, price, id
        host_lat  : float — latitude of the host's apartment
        host_lon  : float — longitude of the host's apartment
    
    Returns:
        list of K prices (float) of the nearest competitor listings
    """

    # ── Extract coordinates into a 2D numpy array ──────────────
    # Shape: (N, 2) where each row is [latitude, longitude]
    # This is the input to the KD-Tree construction
    coordinates = np.array(
        [[listing["latitude"], listing["longitude"]] for listing in listings]
    )

    # ── KD-Tree Construction — O(N log N) ──────────────────────
    # scipy builds a balanced KD-Tree with alternating median cuts
    # on latitude (level 0), longitude (level 1), latitude (level 2)...
    # Tree depth ≈ log₂(N) levels
    tree = KDTree(coordinates)

    # ── Edge Case: small category ───────────────────────────────
    # If N is small (e.g. Shared room in remote neighbourhood),
    # requesting K=50 neighbors on 80 listings returns almost everything.
    # We adapt K dynamically to stay meaningful.
    N = len(listings)
    K = min(50, N // 2)

    print(f"[Stage B] Category size: {N} listings → using K={K} neighbors")

    # ── KNN Query — Θ(log N + K) ───────────────────────────────
    # The tree prunes branches geometrically too far from host_coords.
    # Returns: distances array (unused) and indices into 'listings'
    host_coords = [host_lat, host_lon]
    distances, indices = tree.query(host_coords, k=K)

    # ── Extract prices of K nearest competitors ─────────────────
    neighbor_prices = [listings[i]["price"] for i in indices]

    print(f"[Stage B] Found {K} nearest neighbors")
    print(f"[Stage B] Price range: €{min(neighbor_prices):.0f} — €{max(neighbor_prices):.0f}")

    return neighbor_prices

## Stage C — Statistical Filter: QuickSelect (Median Finding)

## Objective
From the K=50 competitor prices returned by Stage B, we need to find the median
price to suggest to the host. The median is robust to outliers by definition —
a listing at €10,515/night cannot distort it, unlike the arithmetic mean.

## Why the Median and not the Mean?
Our dataset has: min=€8, max=€10,515, std=€443.
The arithmetic mean is €197 — heavily pulled upward by extreme values.
The median (50th percentile) represents the true "market center price"
because exactly half the competitors are above it and half are below.

## Why QuickSelect and not Sorting?
Sorting (MergeSort/HeapSort) would cost O(K log K) to order all 50 prices.
But we do not need the full ordered array — we only need the element at
position K//2. QuickSelect finds exactly that element in Θ(K) expected time,
discarding the irrelevant half at each partition step.

For K=50: QuickSelect ≈ 50 operations vs sorting ≈ 50×log₂(50) ≈ 280 operations.
A 5x improvement — and the gap grows with K.

**Space complexity: Θ(1)** — works in-place, no auxiliary array needed.

**Edge case — even K:**
With K=50 (even number), the true median is the average of the 25th and 26th
elements. We run QuickSelect twice: once for position K//2, once for K//2 - 1,
then average the two values.

In [13]:
import random

# ─────────────────────────────────────────────────────────────
# STAGE C — QuickSelect (Median Finding)
# Time complexity:  Θ(K) expected — single partial partition
# Space complexity: Θ(1) — in-place, no auxiliary array
# ─────────────────────────────────────────────────────────────

def quickselect(prices, target_index):
    """
    Find the element at target_index in the sorted order of prices.
    Works in-place using Hoare's partition scheme.
    
    Parameters:
        prices       : list of floats — competitor prices (modified in-place)
        target_index : int — the rank we are looking for (0-based)
    
    Returns:
        float — the price at target_index in sorted order
    """

    left, right = 0, len(prices) - 1

    while left < right:

        # ── Random pivot selection ──────────────────────────────
        # Avoids worst case Θ(K²) on already-sorted input
        pivot_index = random.randint(left, right)
        pivot = prices[pivot_index]

        # ── Move pivot to end temporarily ───────────────────────
        prices[pivot_index], prices[right] = prices[right], prices[pivot_index]

        # ── Two-pointer partition (Hoare's scheme) ──────────────
        # All elements < pivot go left, all elements > pivot go right
        # This is the same partition logic from QuickSort
        store = left
        for i in range(left, right):
            if prices[i] <= pivot:
                prices[store], prices[i] = prices[i], prices[store]
                store += 1

        # ── Place pivot in its final sorted position ─────────────
        prices[store], prices[right] = prices[right], prices[store]

        # ── Discard the half that cannot contain target_index ────
        # This is what makes QuickSelect faster than full sorting
        if store == target_index:
            return prices[store]
        elif store < target_index:
            left = store + 1    # target is in the right half
        else:
            right = store - 1   # target is in the left half

    return prices[left]


def find_median(neighbor_prices):
    """
    Find the median of K competitor prices using QuickSelect.
    Handles both even and odd K correctly.
    
    Parameters:
        neighbor_prices : list of floats — K competitor prices from Stage B
    
    Returns:
        float — the suggested median price
    """

    prices = neighbor_prices.copy()  # preserve original list
    K = len(prices)

    if K % 2 == 1:
        # Odd K — single median element
        median = quickselect(prices, K // 2)
    else:
        # Even K — average of the two central elements
        # Must copy again because quickselect modifies in-place
        low  = quickselect(prices.copy(), K // 2 - 1)
        high = quickselect(prices.copy(), K // 2)
        median = (low + high) / 2

    print(f"[Stage C] Median price computed over {K} competitors: €{median:.2f}")
    return median

# Main algoritm

In [ ]:
# ─────────────────────────────────────────────────────────────
# MAIN — Airbnb Market Radar
# Connects all stages and handles user input
# ─────────────────────────────────────────────────────────────

def get_user_input():
    """
    Collect host input from terminal.
    Uses numbered menus to prevent typos on categorical variables.
    
    Returns:
        tuple: (room_type, neighbourhood, latitude, longitude)
    """

    print("\n" + "=" * 60)
    print("       AIRBNB MARKET RADAR — Price Suggester")
    print("=" * 60)

    # ── Room Type Selection ─────────────────────────────────────
    room_types = [
        "Entire home/apt",
        "Private room",
        "Hotel room",
        "Shared room"
    ]

    print("\nSelect your room type:")
    for i, rt in enumerate(room_types, 1):
        print(f"  [{i}] {rt}")

    while True:
        try:
            choice = int(input("\nEnter number (1-4): "))
            if 1 <= choice <= 4:
                room_type = room_types[choice - 1]
                break
            else:
                print("Invalid choice. Enter a number between 1 and 4.")
        except ValueError:
            print("Invalid input. Enter a number.")

    # ── Neighbourhood Selection ─────────────────────────────────
    neighbourhoods = [
        "I Centro Storico",
        "II Parioli/Nomentano",
        "III Monte Sacro",
        "IV Tiburtina",
        "V Prenestino/Centocelle",
        "VI Roma delle Torri",
        "VII San Giovanni/Cinecittà",
        "VIII Appia Antica",
        "IX Eur",
        "X Ostia/Acilia",
        "XI Arvalia/Portuense",
        "XII Monte Verde",
        "XIII Aurelia",
        "XIV Monte Mario",
        "XV Cassia/Flaminia"
    ]

    print("\nSelect your neighbourhood:")
    for i, nb in enumerate(neighbourhoods, 1):
        print(f"  [{i:>2}] {nb}")

    while True:
        try:
            choice = int(input("\nEnter number (1-15): "))
            if 1 <= choice <= 15:
                neighbourhood = neighbourhoods[choice - 1]
                break
            else:
                print("Invalid choice. Enter a number between 1 and 15.")
        except ValueError:
            print("Invalid input. Enter a number.")

    # ── Coordinates Input ───────────────────────────────────────
    print("\nEnter your apartment coordinates:")

    while True:
        try:
            latitude = float(input("  Latitude  (e.g. 41.8967): "))
            if 41.7 <= latitude <= 42.1:
                break
            else:
                print("Latitude out of Rome bounds (41.7 - 42.1).")
        except ValueError:
            print("Invalid input. Enter a decimal number.")

    while True:
        try:
            longitude = float(input("  Longitude (e.g. 12.4822): "))
            if 12.2 <= longitude <= 12.8:
                break
            else:
                print("Longitude out of Rome bounds (12.2 - 12.8).")
        except ValueError:
            print("Invalid input. Enter a decimal number.")

    return room_type, neighbourhood, latitude, longitude


def run_pipeline(hash_map, room_type, neighbourhood, latitude, longitude):
    """
    Execute the full 3-stage pipeline for a given host input.
    
    Parameters:
        hash_map     : dict — built in Stage A
        room_type    : str  — host's room type
        neighbourhood: str  — host's neighbourhood
        latitude     : float
        longitude    : float
    
    Returns:
        float — suggested median price
    """

    print("\n" + "=" * 60)
    print("  RUNNING PIPELINE")
    print("=" * 60)

    # ── Stage A — Hash Map Lookup ───────────────────────────────
    key = f"{room_type}_{neighbourhood}"
    print(f"\n[Stage A] Looking up category: '{key}'")

    if key not in hash_map:
        print(f"[Stage A] ERROR: No listings found for '{key}'.")
        print("          Try a different room type or neighbourhood.")
        return None

    listings = hash_map[key]
    print(f"[Stage A] Found {len(listings):,} competitors in this category.")

    # ── Stage B — KD-Tree + KNN ─────────────────────────────────
    neighbor_prices = build_and_query_kdtree(listings, latitude, longitude)

    # ── Stage C — QuickSelect ───────────────────────────────────
    suggested_price = find_median(neighbor_prices)

    return suggested_price


def main():

    # ── Build Hash Map once at startup ──────────────────────────
    # Stage 0 + Stage A run only once regardless of how many
    # queries the host makes — O(N) paid once, Θ(1) per lookup
    print("Loading dataset and building Hash Map...")
    # [Stage 0 + Stage A code here — already written above]
    print(f"[Startup] Hash Map ready: {len(hash_map)} categories loaded.\n")

    # ── Query loop — host can run multiple queries ───────────────
    while True:
        room_type, neighbourhood, latitude, longitude = get_user_input()

        suggested_price = run_pipeline(
            hash_map, room_type, neighbourhood, latitude, longitude
        )

        if suggested_price is not None:
            print("\n" + "=" * 60)
            print(f"  SUGGESTED NIGHTLY PRICE: €{suggested_price:.2f}")
            print("=" * 60)

        # ── Ask if host wants another query ─────────────────────
        again = input("\nRun another query? (y/n): ").strip().lower()
        if again != "y":
            print("\nThank you for using Airbnb Market Radar. Goodbye.")
            break


if __name__ == "__main__":
    main()

Loading dataset and building Hash Map...
[Startup] Hash Map ready: 50 categories loaded.


       AIRBNB MARKET RADAR — Price Suggester

Select your room type:
  [1] Entire home/apt
  [2] Private room
  [3] Hotel room
  [4] Shared room
Invalid input. Enter a number.
Invalid input. Enter a number.
Invalid input. Enter a number.
Invalid input. Enter a number.
Invalid input. Enter a number.
Invalid input. Enter a number.
